File: harness/qmp.py
"""Minimal synchronous client for QEMU's machine protocol."""

from __future__ import annotations

import json
import socket
import time
from os import PathLike
from typing import TextIO

_TIMEOUT_SECONDS = 5.0


class QmpError(RuntimeError):
    """A QMP protocol or command error."""


class QmpClient:
    """Control one QEMU instance through a Unix-domain QMP socket."""

    def __init__(self, socket_path: str | PathLike[str]) -> None:
        self._socket_path = str(socket_path)
        self._socket: socket.socket | None = None
        self._stream: TextIO | None = None
        self._next_id = 1

    def connect(self) -> None:
        """Connect, validate QEMU's greeting, and negotiate QMP capabilities."""
        if self._socket is not None:
            raise QmpError("QMP client is already connected")

        deadline = time.monotonic() + _TIMEOUT_SECONDS
        while True:
            connection = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
            try:
                connection.connect(self._socket_path)
            except (FileNotFoundError, ConnectionRefusedError) as error:
                connection.close()
                remaining = deadline - time.monotonic()
                if remaining <= 0:
                    raise QmpError(
                        f"timed out connecting to QMP socket {self._socket_path}"
                    ) from error
                time.sleep(min(0.01, remaining))
                continue
            except BaseException:
                connection.close()
                raise

            connection.settimeout(_TIMEOUT_SECONDS)
            self._socket = connection
            try:
                self._stream = connection.makefile("rw", encoding="utf-8", newline="\n")
                self._validate_greeting(self._read_message())
                self._execute("qmp_capabilities")
            except BaseException:
                self.close()
                raise
            return

    def query_status(self) -> str:
        result = self._execute("query-status")
        if not isinstance(result, dict) or not isinstance(result.get("status"), str):
            raise QmpError("query-status returned an invalid response")
        return result["status"]

    def stop(self) -> None:
        self._execute("stop")

    def cont(self) -> None:
        self._execute("cont")

    def quit(self) -> None:
        self._execute("quit")

    def close(self) -> None:
        stream, connection = self._stream, self._socket
        self._stream = None
        self._socket = None
        try:
            if stream is not None:
                stream.close()
        finally:
            if connection is not None:
                connection.close()

    def __enter__(self) -> QmpClient:
        self.connect()
        return self

    def __exit__(self, *exc_info: object) -> None:
        self.close()

    @staticmethod
    def _validate_greeting(greeting: dict[str, object]) -> None:
        qmp = greeting.get("QMP")
        if (
            not isinstance(qmp, dict)
            or not isinstance(qmp.get("version"), dict)
            or not isinstance(qmp.get("capabilities"), list)
        ):
            raise QmpError("invalid QMP greeting")

    def _execute(self, command: str) -> object:
        if self._stream is None:
            raise QmpError("QMP client is not connected")

        request_id = self._next_id
        self._next_id += 1
        request = {"execute": command, "id": request_id}
        self._stream.write(json.dumps(request, separators=(",", ":")) + "\r\n")
        self._stream.flush()

        while True:
            response = self._read_message()
            if "event" in response:
                continue
            if response.get("id") != request_id:
                raise QmpError(
                    f"{command} received response for unexpected id {response.get('id')!r}"
                )

            error = response.get("error")
            if error is not None:
                if not isinstance(error, dict):
                    raise QmpError(f"{command} returned an invalid error response")
                error_class = error.get("class", "unknown error")
                description = error.get("desc", "no description")
                raise QmpError(f"{command} failed: {error_class}: {description}")
            if "return" not in response:
                raise QmpError(f"{command} returned no result")
            return response["return"]

    def _read_message(self) -> dict[str, object]:
        if self._stream is None:
            raise QmpError("QMP client is not connected")

        line = self._stream.readline()
        if not line:
            raise QmpError("QMP connection closed")
        try:
            message = json.loads(line)
        except json.JSONDecodeError as error:
            raise QmpError("QMP returned invalid JSON") from error
        if not isinstance(message, dict):
            raise QmpError("QMP returned a non-object message")
        return message

File: internal/qemu/qmp.go
package qemu

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net"
	"os"
	"strconv"
	"syscall"
	"time"
	"unicode/utf8"
)

const (
	qmpTimeout        = 5 * time.Second
	maxQMPMessageSize = 1024 * 1024
)

type QMPError struct {
	Reason string
	Err    error
}

func (e *QMPError) Error() string {
	if e.Err == nil {
		return e.Reason
	}
	return e.Reason + ": " + e.Err.Error()
}

func (e *QMPError) Unwrap() error { return e.Err }

// QMPClient synchronously controls one QEMU process. One caller owns the
// client; it has no background read loop.
type QMPClient struct {
	socketPath string
	timeout    time.Duration
	connection net.Conn
	reader     *bufio.Reader
	writer     *bufio.Writer
	nextID     uint64
}

func NewQMPClient(socketPath string) *QMPClient {
	return &QMPClient{socketPath: socketPath, timeout: qmpTimeout, nextID: 1}
}

func (c *QMPClient) Connect(ctx context.Context) error {
	if c.connection != nil {
		return &QMPError{Reason: "QMP client is already connected"}
	}
	deadline := time.Now().Add(c.timeout)
	if contextDeadline, ok := ctx.Deadline(); ok && contextDeadline.Before(deadline) {
		deadline = contextDeadline
	}

	for {
		if err := ctx.Err(); err != nil {
			return err
		}
		dialer := net.Dialer{Deadline: deadline}
		connection, err := dialer.DialContext(ctx, "unix", c.socketPath)
		if err == nil {
			c.connection = connection
			c.reader = bufio.NewReader(connection)
			c.writer = bufio.NewWriter(connection)
			stopCancellation := interruptConnectionOnCancel(ctx, connection)
			if err := c.readGreeting(ctx); err != nil {
				stopCancellation()
				c.Close()
				return err
			}
			if _, err := c.execute(ctx, "qmp_capabilities"); err != nil {
				stopCancellation()
				c.Close()
				return err
			}
			stopCancellation()
			return nil
		}
		if contextErr := ctx.Err(); contextErr != nil {
			return contextErr
		}
		if time.Until(deadline) <= 0 {
			return &QMPError{Reason: fmt.Sprintf("timed out connecting to QMP socket %s", c.socketPath), Err: err}
		}
		if !retryableQMPConnectError(err) {
			return err
		}
		remaining := time.Until(deadline)
		if remaining <= 0 {
			if contextErr := ctx.Err(); contextErr != nil {
				return contextErr
			}
			return &QMPError{Reason: fmt.Sprintf("timed out connecting to QMP socket %s", c.socketPath), Err: err}
		}
		wait := min(10*time.Millisecond, remaining)
		timer := time.NewTimer(wait)
		select {
		case <-ctx.Done():
			if !timer.Stop() {
				<-timer.C
			}
			return ctx.Err()
		case <-timer.C:
		}
	}
}

func (c *QMPClient) QueryStatus(ctx context.Context) (string, error) {
	result, err := c.execute(ctx, "query-status")
	if err != nil {
		return "", err
	}
	object, ok := result.(map[string]any)
	if !ok {
		return "", &QMPError{Reason: "query-status returned an invalid response"}
	}
	status, ok := object["status"].(string)
	if !ok {
		return "", &QMPError{Reason: "query-status returned an invalid response"}
	}
	return status, nil
}

func (c *QMPClient) Stop(ctx context.Context) error {
	_, err := c.execute(ctx, "stop")
	return err
}

func (c *QMPClient) Continue(ctx context.Context) error {
	_, err := c.execute(ctx, "cont")
	return err
}

func (c *QMPClient) Quit(ctx context.Context) error {
	_, err := c.execute(ctx, "quit")
	return err
}

func (c *QMPClient) Close() error {
	connection := c.connection
	c.connection = nil
	c.reader = nil
	c.writer = nil
	if connection == nil {
		return nil
	}
	return connection.Close()
}

func (c *QMPClient) readGreeting(ctx context.Context) error {
	greeting, err := c.readMessage(ctx)
	if err != nil {
		return err
	}
	qmp, ok := greeting["QMP"].(map[string]any)
	if !ok {
		return &QMPError{Reason: "invalid QMP greeting"}
	}
	if _, ok := qmp["version"].(map[string]any); !ok {
		return &QMPError{Reason: "invalid QMP greeting"}
	}
	if _, ok := qmp["capabilities"].([]any); !ok {
		return &QMPError{Reason: "invalid QMP greeting"}
	}
	return nil
}

func (c *QMPClient) execute(ctx context.Context, command string) (any, error) {
	if c.connection == nil || c.writer == nil {
		return nil, &QMPError{Reason: "QMP client is not connected"}
	}
	stopCancellation := interruptConnectionOnCancel(ctx, c.connection)
	defer stopCancellation()
	requestID := c.nextID
	if requestID == 0 {
		return nil, &QMPError{Reason: "QMP request ID space is exhausted"}
	}
	c.nextID++
	request := struct {
		Execute string `json:"execute"`
		ID      uint64 `json:"id"`
	}{Execute: command, ID: requestID}
	encoded, err := json.Marshal(request)
	if err != nil {
		return nil, &QMPError{Reason: command + " request could not be encoded", Err: err}
	}
	encoded = append(encoded, '\r', '\n')
	if err := c.setWriteDeadline(ctx); err != nil {
		return nil, err
	}
	if _, err := c.writer.Write(encoded); err != nil {
		return c.closeWithError(qmpOperationError(ctx, command+" request failed", err))
	}
	if err := c.writer.Flush(); err != nil {
		return c.closeWithError(qmpOperationError(ctx, command+" request failed", err))
	}

	for {
		response, err := c.readMessage(ctx)
		if err != nil {
			return c.closeWithError(err)
		}
		if _, event := response["event"]; event {
			continue
		}
		if !qmpResponseIDMatches(response["id"], requestID) {
			return c.closeWithError(&QMPError{Reason: fmt.Sprintf("%s received response for unexpected id %v", command, response["id"])})
		}
		if responseError, exists := response["error"]; exists && responseError != nil {
			object, ok := responseError.(map[string]any)
			if !ok {
				return c.closeWithError(&QMPError{Reason: command + " returned an invalid error response"})
			}
			errorClass, exists := object["class"]
			if !exists {
				errorClass = "unknown error"
			}
			description, exists := object["desc"]
			if !exists {
				description = "no description"
			}
			return nil, &QMPError{Reason: fmt.Sprintf("%s failed: %v: %v", command, errorClass, description)}
		}
		result, exists := response["return"]
		if !exists {
			return c.closeWithError(&QMPError{Reason: command + " returned no result"})
		}
		return result, nil
	}
}

func (c *QMPClient) closeWithError(err error) (any, error) {
	_ = c.Close()
	return nil, err
}

func (c *QMPClient) readMessage(ctx context.Context) (map[string]any, error) {
	if c.connection == nil || c.reader == nil {
		return nil, &QMPError{Reason: "QMP client is not connected"}
	}
	line, err := c.readLine(ctx)
	if err != nil {
		return nil, err
	}
	if !utf8.Valid(line) {
		return nil, &QMPError{Reason: "QMP returned invalid JSON"}
	}
	decoder := json.NewDecoder(bytes.NewReader(line))
	decoder.UseNumber()
	var message any
	if err := decoder.Decode(&message); err != nil {
		return nil, &QMPError{Reason: "QMP returned invalid JSON", Err: err}
	}
	var trailing any
	if err := decoder.Decode(&trailing); !errors.Is(err, io.EOF) {
		return nil, &QMPError{Reason: "QMP returned invalid JSON", Err: err}
	}
	object, ok := message.(map[string]any)
	if !ok {
		return nil, &QMPError{Reason: "QMP returned a non-object message"}
	}
	return object, nil
}

func (c *QMPClient) readLine(ctx context.Context) ([]byte, error) {
	line := make([]byte, 0, 4096)
	for {
		if err := c.setReadDeadline(ctx); err != nil {
			return nil, err
		}
		fragment, err := c.reader.ReadSlice('\n')
		if len(fragment) > maxQMPMessageSize-len(line) {
			return nil, &QMPError{Reason: "QMP message exceeds 1 MiB"}
		}
		line = append(line, fragment...)
		switch {
		case err == nil:
			return line, nil
		case errors.Is(err, bufio.ErrBufferFull):
			continue
		case errors.Is(err, io.EOF) && len(line) > 0:
			return line, nil
		case errors.Is(err, os.ErrDeadlineExceeded) || errors.Is(err, context.DeadlineExceeded):
			return nil, qmpOperationError(ctx, "QMP read timed out", err)
		case len(line) == 0:
			return nil, &QMPError{Reason: "QMP connection closed", Err: err}
		default:
			return nil, &QMPError{Reason: "QMP connection closed while reading a message", Err: err}
		}
	}
}

func (c *QMPClient) setReadDeadline(ctx context.Context) error {
	deadline, err := c.operationDeadline(ctx)
	if err != nil {
		return err
	}
	if err := c.connection.SetReadDeadline(deadline); err != nil {
		return &QMPError{Reason: "could not set QMP read deadline", Err: err}
	}
	return nil
}

func (c *QMPClient) setWriteDeadline(ctx context.Context) error {
	deadline, err := c.operationDeadline(ctx)
	if err != nil {
		return err
	}
	if err := c.connection.SetWriteDeadline(deadline); err != nil {
		return &QMPError{Reason: "could not set QMP write deadline", Err: err}
	}
	return nil
}

func (c *QMPClient) operationDeadline(ctx context.Context) (time.Time, error) {
	if err := ctx.Err(); err != nil {
		return time.Time{}, err
	}
	deadline := time.Now().Add(c.timeout)
	if contextDeadline, ok := ctx.Deadline(); ok && contextDeadline.Before(deadline) {
		deadline = contextDeadline
	}
	return deadline, nil
}

func retryableQMPConnectError(err error) bool {
	return errors.Is(err, os.ErrNotExist) || errors.Is(err, syscall.ECONNREFUSED)
}

func qmpResponseIDMatches(value any, expected uint64) bool {
	number, ok := value.(json.Number)
	if !ok {
		return false
	}
	identifier, err := strconv.ParseUint(string(number), 10, 64)
	return err == nil && identifier == expected
}

func qmpOperationError(ctx context.Context, reason string, err error) error {
	if contextErr := ctx.Err(); contextErr != nil {
		return contextErr
	}
	return &QMPError{Reason: reason, Err: err}
}

func interruptConnectionOnCancel(ctx context.Context, connection net.Conn) func() {
	finished := make(chan struct{})
	stop := context.AfterFunc(ctx, func() {
		_ = connection.SetDeadline(time.Now())
		close(finished)
	})
	return func() {
		if !stop() {
			<-finished
		}
	}
}
